### PR TITLE
Enhancing dotenv function

### PR DESCRIPTION
**Pull request by rrmistry**
Enhancing the `dotenv` function with few enhancements:

* Handling environment variable names and values with quotes (`'` or `"`)
* Handling comments in ".env" variable file
* Gracefully moving on if ".env" file name not found
* Printing line, file name, and line number for invalid formatted lines in ".env" files

This should resolve https://github.com/tilt-dev/tilt-extensions/issues/444

### CODE REVIEW

File: dotenv/Tiltfile [hunk 1]
```diff
@@ -1,8 +1,46 @@
...
+            if len(line) > 0:
+                fail("Invalid format for dotenv file: '%s' in '%s:%s'" % (line, fn, lineNumber+1))
+            continue # skip empty lines
+
+        # Print if key is valid
+        print("Loading environment variable '%s'" % v[0])
+
+        # Set environment variable
+        os.putenv(
+            v[0].strip().strip('"').strip("'"),
```
Comment (nicks) [inline dotenv/Tiltfile:44]: i think this will take out all the quotes anywhere in the string, which is not what we want?

File: dotenv/Tiltfile [hunk 2]
```diff
@@ -1,8 +1,46 @@
...
+        )
+
+        # Check if key and value are present
+        if len(v) < 2:
+            if len(line) > 0:
+                fail("Invalid format for dotenv file: '%s' in '%s:%s'" % (line, fn, lineNumber+1))
+            continue # skip empty lines
+
+        # Print if key is valid
+        print("Loading environment variable '%s'" % v[0])
```
Comment (nicks) [inline dotenv/Tiltfile:40]: this seems like it'd be very chatty?

File: dotenv/Tiltfile [hunk 3]
```diff
@@ -1,8 +1,46 @@
...
-			continue # skip empty lines
-		os.putenv(v[0], v[1])
+def dotenv(fn=".env"):
+
+    # Check if file exists
+    if not os.path.exists(fn):
+        warn("Environment variables file not found: '%s'" % fn)
+        return
+
+    print("Loading environment variables from file: '%s'" % fn)
```
Comment (nicks) [inline dotenv/Tiltfile:8]: this seems like an odd thing to print?

**Review by nicks — COMMENTED** [review]
cc @milas , who had a lot of grief over Compose's env file parser. curious what he thinks of this. i realize this extension isn't as widely used as Compose's parser, and so the trade-offs might be different.

File: dotenv/Tiltfile (hunk 3 above)
Comment (rrmistry) [reply dotenv/Tiltfile:8]: The intention here was to provide visibility into which dotenv file is currently being loaded.

Our use case is to have a Git Mono Repo that contains Git Sub-Repos for each of the micro services. And we dynamically load the Titlfile from the sub-repos via the Tiltfile in the Mono Repo.

IMHO, this is still good to have. But if this seems to verbose then I can think of adding a `verbose` or `debug` parameter in the parent function that would provide this diagnostic information.

What are your thoughts?

```suggestion
    def dotenv(fn=".env", verbose=False):
    # ...
    if verbose:
        print("Loading environment variables from file: '%s'" % fn)
```

File: dotenv/Tiltfile (hunk 2 above)
Comment (rrmistry) [reply dotenv/Tiltfile:40]: Same as the previous comment, I can think of adding a `verbose` or `debug` parameter in the parent function that would provide this diagnostic information.

What are your thoughts?

```suggestion
    def dotenv(fn=".env", verbose=False):
    # ...
    for lineNumber in range(len(lines)):
        # ...
        if verbose:
            print("Loading environment variable '%s'" % v[0])
        # ...
```

File: dotenv/Tiltfile (hunk 1 above)
Comment (rrmistry) [reply dotenv/Tiltfile]: Thanks @nicks , good callout on the quotes

This is now fixed in https://github.com/tilt-dev/tilt-extensions/pull/498/commits/b64bec7bc92b7d6e12d2dd317b9617a784660973 and ready for review

**Comment by atta** [comment]
Hi a nice thing to add would be: to also remove the \r on the end of each line if it exist for better windows support

**Comment by rrmistry** [comment]
> Hi a nice thing to add would be: to also remove the \r on the end of each line if it exist for better windows support

The current PR relies on the StarLark function [`strip`](https://bazel.build/rules/lib/core/string#strip).

Looking at the documentation, it seems that when called without any parameters then it removes all whitespace. But it is not explicitly clear if it is `\n` or `\r` or both.

A simple task could be to add an explicit test case for this to make sure future breaking changes are caught (if any) for `\r` character support.

I can update this PR as soon as I get a chance. Thanks for pointing this out @atta !

**Pull request merged**